aten/src/ATen/native/Linear.cpp

```diff
@@ -383,144 +393,146 @@ Tensor einsum(c10::string_view equation, TensorList operands) {
-              << permuted_operands[j].sizes();
-        }
-        TORCH_CHECK(false, msg.str());
-      }
-      if (dim_size != 1) {
-        broadcast_size = dim_size;
-        dim_last_op[dim] = i;
+    for (auto& val : permutation) {
+      if (val == -1) {
+        op = op.unsqueeze(dim);
+        val = dim++;
       }
     }
-    has_zero_size_dim |= broadcast_size == 0;
+    ops.emplace_back(op.permute(permutation));
   }
 
-  // Compute result
-  Tensor result = permuted_operands[0];
+  const auto path = optimize.value_or(std::vector<int64_t>{});
+  auto it = path.begin();
 
-  // Fast path for when an operand has zero sized dim
-  if (has_zero_size_dim) {
-    std::vector<int64_t> out_shape(out_size);
-    for (const auto i : c10::irange(out_size)) {
-      out_shape[i] = permuted_operands[dim_last_op[i]].size(i);
-    }
-    return at::zeros(out_shape, result.options());
-  }
+  // Contract
+  while (ops.size() > 1) {
+    int64_t i = 0;
+    int64_t j = 1;
 
-  // Sum out or squeeze dimensions that are size 1 for all later operands
-  int64_t dim = out_size;
-  for (int64_t i = dim; i < perm_index; ++i, ++dim) {
-    if (dim_last_op[i] == 0) {
-      if (result.size(dim) == 1) {
-        result = result.squeeze(dim--);
-      } else {
-        result = result.sum(dim--);
+    if (optimize.has_value()) {
+      i = *it++;
+      j = *it++;
+      if (j < i) {
+        std::swap(i, j);
       }
+
+      TORCH_CHECK(
+          i != j && i >= 0 && j < static_cast<int64_t>(ops.size()),
+          "einsum(): invalid contraction (",
+          i,
+          ", ",
+          j,
+          i == j ? ") cannot contract an operand with itself"
+                 : ") operand index is out of bounds");
     }
-  }
 
-  for (const auto i: c10::irange(1, num_ops)) {
-    Tensor operand = permuted_operands[i];
-    std::vector<int64_t> sum_dims;
+    auto a = ops[i];
+    auto b = ops[j];
+    ops.erase(ops.begin() + j);
+    ops.erase(ops.begin() + i);
 
-    // Sum out or squeeze dimensions that are size 1 for all later operands
-    dim = out_size;
-    for (int64_t j = dim; j < perm_index; ++j, ++dim) {
-      if (dim_last_op[j] < i) {
-        operand = operand.squeeze(dim);
-        --dim;
-      } else if (dim_last_op[j] == i) {
-        if (result.size(dim) == 1) {
-          operand = operand.sum(dim);
-          result = result.squeeze(dim);
-          --dim;
-        } else {
+    // Collect dimensions that can be summed now
+    std::vector<int64_t> sum_dims;
+    for (auto dim = out_num_dim; dim < perm_index; ++dim) {
+      if (a.size(dim) != 1 && b.size(dim) != 1) {
+        if (--dim_counts[dim] == 1) {
           sum_dims.push_back(dim);
+          dim_counts[dim] = 0;
+        }
+      } else if (dim_counts[dim] == 1) {
+        if (a.size(dim) != 1) {
+          a = a.sum(dim, true);
+          dim_counts[dim] = 0;
+        } else if (b.size(dim) != 1) {
+          b = b.sum(dim, true);
+          dim_counts[dim] = 0;
```

Comment:
Isn't it possible to sum all the dimensions from `a` and `b` outside the loop just calling just one `sum` kernel? This would be faster and marginally more stable.

Reply:
Conclusion to this: I've added the delay in summing + used smallvector to make it not incur too much overhead